src/devices/src/virtio/queue.rs

```diff
@@ -789,15 +819,29 @@ pub(crate) mod tests {
         let mut q = vq.create_queue();
         assert_eq!(vq.used.idx.get(), 0);
 
-        //index too large
-        q.add_used(m, 16, 0x1000);
-        assert_eq!(vq.used.idx.get(), 0);
+        //Valid queue addresses configuration
+        {
+            //index too large
+            q.add_used(m, 16, 0x1000).unwrap();
+            assert_eq!(vq.used.idx.get(), 0);
 
-        //should be ok
-        q.add_used(m, 1, 0x1000);
-        assert_eq!(vq.used.idx.get(), 1);
-        let x = vq.used.ring[0].get();
-        assert_eq!(x.id, 1);
-        assert_eq!(x.len, 0x1000);
+            //should be ok
+            q.add_used(m, 1, 0x1000).unwrap();
+            assert_eq!(vq.used.idx.get(), 1);
+            let x = vq.used.ring[0].get();
+            assert_eq!(x.id, 1);
+            assert_eq!(x.len, 0x1000);
+        }
+
+        //Invalid queue addresses configuration
+        {
+            q.used_ring = GuestAddress(0xffff_ffff);
+
+            //should fail
+            assert!(match q.add_used(m, 1, 0x1000) {
+                Err(QueueError::GuestMemoryError(_)) => true,
+                _ => false,
+            });
```

Comment:
Since both `QueueError` and `GuestMemoryError` implement Display, I would test `add_used()` errors like this:
```rust
assert_eq!(format!("{}", q.add_used(m, 1, 0x1000).unwrap_err()), format!("Failed to write value into the virtio queue used ring: Guest memory error: invalid guest address {}", invalid_addr));
```
and I would somehow compute `invalid_addr` based on the other addresses / offsets.
This way, the testing is more precise and you also test the Display functionality that you implemented for `QueueError`.

Also, just curious: is it possible to test the other two `QueueError` scenarios too?

Reply:
We decided a while ago to not duplicate implementation in tests. @andreeaflorescu had a nice article around it.

Tests that were matching exact error messages were deemed unuseful. Matching string literal duplicated both in impl and tests is redundant and not really useful.

We decided to test `Debug` and `Display` by just doing `format!("{:?}{}", err, err)` which verifies (at test compile time) that `Debug` and `Display` are implemented.

Reply:
Ok, sounds good too.